File: frontend/src/app/core/services/role.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, catchError, throwError } from 'rxjs';
import { environment } from '../../../environments/environment';
import { 
  Role, 
  CreateRoleRequest, 
  UpdateRoleRequest, 
  RoleFilters 
} from '../models/role.model';
import { PaginatedResponse } from '../models/pagination.model';

/**
 * Servicio para gestión de roles
 * CRUD completo de roles del sistema
 */
@Injectable({
  providedIn: 'root'
})
export class RoleService {

  private readonly API_URL = `${environment.apiUrl}/roles`;

  constructor(private http: HttpClient) {}

  // ========== LISTAR ROLES ==========

  /**
   * Obtiene lista paginada de roles con filtros
   */
  getRoles(filters?: RoleFilters): Observable<PaginatedResponse<Role>> {
    let params = new HttpParams();

    if (filters) {
      if (filters.search) params = params.set('search', filters.search);
      if (filters.status !== undefined) params = params.set('status', filters.status.toString());
      if (filters.page !== undefined) params = params.set('page', filters.page.toString());
      if (filters.size !== undefined) params = params.set('size', filters.size.toString());
      if (filters.sortBy) params = params.set('sortBy', filters.sortBy);
      if (filters.sortDirection) params = params.set('sortDirection', filters.sortDirection);
    }

    return this.http.get<PaginatedResponse<Role>>(this.API_URL, { params }).pipe(
      catchError(this.handleError)
    );
  }

  /**
   * Obtiene todos los roles sin paginación (para selects)
   */
  getAllRoles(): Observable<Role[]> {
    return this.http.get<Role[]>(`${this.API_URL}/all`).pipe(
      catchError(this.handleError)
    );
  }

  // ========== OBTENER ROL POR ID ==========

  /**
   * Obtiene un rol por su ID
   */
  getRoleById(id: number): Observable<Role> {
    return this.http.get<Role>(`${this.API_URL}/${id}`).pipe(
      catchError(this.handleError)
    );
  }

  // ========== CREAR ROL ==========

  /**
   * Crea un nuevo rol
   */
  createRole(request: CreateRoleRequest): Observable<Role> {
    return this.http.post<Role>(this.API_URL, request).pipe(
      catchError(this.handleError)
    );
  }

  // ========== ACTUALIZAR ROL ==========

  /**
   * Actualiza un rol existente
   */
  updateRole(id: number, request: UpdateRoleRequest): Observable<Role> {
    return this.http.put<Role>(`${this.API_URL}/${id}`, request).pipe(
      catchError(this.handleError)
    );
  }

  // ========== ELIMINAR ROL ==========

  /**
   * Elimina un rol (solo si no tiene usuarios asignados)
   */
  deleteRole(id: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${id}`).pipe(
      catchError(this.handleError)
    );
  }

  // ========== GESTIÓN DE PERMISOS ==========

  /**
   * Asigna permisos a un rol
   */
  assignPermissions(roleId: number, permissionIds: number[]): Observable<Role> {
    return this.http.post<Role>(`${this.API_URL}/${roleId}/permissions`, { permissionIds }).pipe(
      catchError(this.handleError)
    );
  }

  /**
   * Remueve un permiso de un rol
   */
  removePermission(roleId: number, permissionId: number): Observable<Role> {
    return this.http.delete<Role>(`${this.API_URL}/${roleId}/permissions/${permissionId}`).pipe(
      catchError(this.handleError)
    );
  }

  // ========== MANEJO DE ERRORES ==========

  private handleError(error: any): Observable<never> {
    console.error('❌ Error en RoleService:', error);
    return throwError(() => error);
  }
}

File: backend/services/auth-service/src/main/java/com/winnersystems/smartparking/auth/application/service/role/RoleService.java
package com.winnersystems.smartparking.auth.application.service.role;

import com.winnersystems.smartparking.auth.application.dto.query.PagedResponse;
import com.winnersystems.smartparking.auth.application.dto.query.PermissionDto;
import com.winnersystems.smartparking.auth.application.dto.query.RoleDto;
import com.winnersystems.smartparking.auth.application.port.output.PermissionPersistencePort;
import com.winnersystems.smartparking.auth.application.port.output.RolePersistencePort;
import com.winnersystems.smartparking.auth.domain.model.Permission;
import com.winnersystems.smartparking.auth.domain.model.Role;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * Servicio de aplicación para gestión de roles.
 *
 * @author Edwin Yoner Winner Systems - Smart Parking Platform
 * @version 1.0
 */
@Service
@RequiredArgsConstructor
@Transactional
public class RoleService {

   private final RolePersistencePort rolePersistencePort;
   private final PermissionPersistencePort permissionPersistencePort;

   /**
    * Lista roles con paginación y filtros
    */
   public PagedResponse<RoleDto> listRoles(String search, Boolean status, int page, int size) {
      // Obtener todos los roles activos
      List<Role> roles = rolePersistencePort.findAllActive();

      // Filtrar por búsqueda si existe
      if (search != null && !search.isBlank()) {
         String searchLower = search.toLowerCase();
         roles = roles.stream()
               .filter(role ->
                     role.getName().toLowerCase().contains(searchLower) ||
                           (role.getDescription() != null && role.getDescription().toLowerCase().contains(searchLower))
               )
               .collect(Collectors.toList());
      }

      // Filtrar por estado si existe
      if (status != null) {
         roles = roles.stream()
               .filter(role -> role.isActive() == status)
               .collect(Collectors.toList());
      }

      // Calcular paginación manual
      int start = page * size;
      int end = Math.min(start + size, roles.size());

      // Validar que start no sea mayor que el tamaño de la lista
      if (start >= roles.size()) {
         return PagedResponse.of(List.of(), page, size, roles.size());
      }

      List<Role> paginatedRoles = roles.subList(start, end);

      // Mapear a DTOs
      List<RoleDto> roleDtos = paginatedRoles.stream()
            .map(this::mapToDto)
            .collect(Collectors.toList());

      // Crear respuesta paginada usando el método estático
      return PagedResponse.of(roleDtos, page, size, roles.size());
   }

   /**
    * Obtiene un rol por ID
    */
   public RoleDto getRoleById(Long id) {
      Role role = rolePersistencePort.findById(id)
            .orElseThrow(() -> new RuntimeException("Rol no encontrado: " + id));
      return mapToDto(role);
   }

   /**
    * Crea un nuevo rol
    */
   public RoleDto createRole(String name, String description, boolean status, Set<Long> permissionIds) {
      // Cargar permisos
      Set<Permission> permissions = new HashSet<>();
      if (permissionIds != null && !permissionIds.isEmpty()) {
         for (Long permissionId : permissionIds) {
            permissionPersistencePort.findById(permissionId)
                  .ifPresent(permissions::add);
         }
      }

      // Crear rol
      Role role = new Role(name, description);
      if (status) {
         role.activate();
      } else {
         role.deactivate();
      }

      // Asignar permisos
      permissions.forEach(role::addPermission);

      // Guardar
      Role savedRole = rolePersistencePort.save(role);
      return mapToDto(savedRole);
   }

   /**
    * Actualiza un rol existente
    */
   public RoleDto updateRole(Long id, String name, String description, Boolean status, Set<Long> permissionIds) {
      Role role = rolePersistencePort.findById(id)
            .orElseThrow(() -> new RuntimeException("Rol no encontrado: " + id));

      // Actualizar campos
      if (name != null && description != null && !role.isSystemRole()) {
         role.updateDetails(name, description);
      }

      if (status != null) {
         if (status) {
            role.activate();
         } else {
            role.deactivate();
         }
      }

      // Actualizar permisos si se proporcionaron
      if (permissionIds != null) {
         role.clearPermissions();
         Set<Permission> permissions = new HashSet<>();
         for (Long permissionId : permissionIds) {
            permissionPersistencePort.findById(permissionId)
                  .ifPresent(permissions::add);
         }
         permissions.forEach(role::addPermission);
      }

      Role updatedRole = rolePersistencePort.save(role);
      return mapToDto(updatedRole);
   }

   /**
    * Elimina un rol
    */
   public void deleteRole(Long id) {
      Role role = rolePersistencePort.findById(id)
            .orElseThrow(() -> new RuntimeException("Rol no encontrado: " + id));

      if (role.isSystemRole()) {
         throw new RuntimeException("No se pueden eliminar roles de sistema");
      }

      role.markAsDeleted(null); // TODO: Obtener ID del usuario autenticado
      rolePersistencePort.save(role);
   }

   /**
    * Mapea Role (domain) a RoleDto
    */
   private RoleDto mapToDto(Role role) {
      Set<PermissionDto> permissionDtos = role.getPermissions().stream()
            .map(p -> new PermissionDto(
                  p.getId(),
                  p.getName(),
                  p.getDescription(),
                  p.isActive(),
                  p.getCreatedAt(),
                  p.getCreatedBy(),
                  p.getUpdatedAt(),
                  p.getUpdatedBy(),
                  p.getDeletedAt(),
                  p.getDeletedBy()
            ))
            .collect(Collectors.toSet());

      return new RoleDto(
            role.getId(),
            role.getName(),
            role.getDescription(),
            role.isActive(),
            permissionDtos,
            role.getCreatedAt(),
            role.getCreatedBy(),
            role.getUpdatedAt(),
            role.getUpdatedBy(),
            role.getDeletedAt(),
            role.getDeletedBy()
      );
   }

   /**
    * Obtiene todos los roles activos sin paginación
    * Usado para selects en formularios
    */
   public List<RoleDto> getAllActiveRoles() {
      List<Role> roles = rolePersistencePort.findAllActive();

      return roles.stream()
            .map(this::mapToDto)
            .collect(Collectors.toList());
   }
}